File: routes/services.js
const express = require('express');
const router = express.Router();
const serviceRegistry = require('../services/serviceRegistry');

/**
 * Get all available services and their connection status
 */
router.get('/', (req, res) => {
  try {
    const availableServices = serviceRegistry.getAvailableServices();
    const connectionStatus = serviceRegistry.getConnectionStatus(req.session);
    const connectedUsers = serviceRegistry.getConnectedServiceUsers(req.session);

    const services = availableServices.map(service => ({
      ...service,
      connected: connectionStatus[service.type] || false,
      user: connectedUsers[service.type] || null,
      capabilities: serviceRegistry.getServiceCapabilities(service.type)
    }));

    res.json({
      services,
      totalConnected: Object.values(connectionStatus).filter(Boolean).length
    });
  } catch (error) {
    console.error('Error getting services:', error);
    res.status(500).json({ error: 'Failed to get services' });
  }
});

/**
 * Get connection status for all services
 */
router.get('/status', (req, res) => {
  try {
    const connectionStatus = serviceRegistry.getConnectionStatus(req.session);
    const connectedUsers = serviceRegistry.getConnectedServiceUsers(req.session);

    res.json({
      status: connectionStatus,
      users: connectedUsers
    });
  } catch (error) {
    console.error('Error getting service status:', error);
    res.status(500).json({ error: 'Failed to get service status' });
  }
});

/**
 * Disconnect a specific service
 */
router.post('/:serviceType/disconnect', (req, res) => {
  try {
    const { serviceType } = req.params;
    
    if (!serviceRegistry.serviceTypes[serviceType]) {
      return res.status(400).json({ error: 'Invalid service type' });
    }

    serviceRegistry.disconnectService(req.session, serviceType);
    
    res.json({ 
      success: true, 
      message: `${serviceRegistry.serviceTypes[serviceType].name} disconnected` 
    });
  } catch (error) {
    console.error('Error disconnecting service:', error);
    res.status(500).json({ error: 'Failed to disconnect service' });
  }
});

/**
 * Get playlists for a specific service
 */
router.get('/:serviceType/playlists', async (req, res) => {
  try {
    const { serviceType } = req.params;
    
    if (!serviceRegistry.isServiceConnected(req.session, serviceType)) {
      return res.status(401).json({ error: `${serviceType} not connected` });
    }

    const connectedServices = serviceRegistry.getConnectedServices(req.session);
    const service = connectedServices[serviceType];

    if (!service) {
      return res.status(404).json({ error: 'Service not found' });
    }

    const playlists = await service.getUserPlaylists();
    
    res.json({
      service: serviceType,
      playlists: playlists.map(playlist => ({
        id: playlist.id,
        name: playlist.name,
        description: playlist.description,
        trackCount: playlist.trackCount,
        image: playlist.image,
        owner: playlist.owner
      }))
    });
  } catch (error) {
    console.error(`Error fetching ${req.params.serviceType} playlists:`, error);
    
    if (error.message.includes('token expired') || error.message.includes('authentication')) {
      return res.status(401).json({ error: 'Authentication expired', code: 'TOKEN_EXPIRED' });
    }
    
    res.status(500).json({ error: 'Failed to fetch playlists' });
  }
});

/**
 * Get tracks for a specific playlist from a specific service
 */
router.get('/:serviceType/playlists/:playlistId/tracks', async (req, res) => {
  try {
    const { serviceType, playlistId } = req.params;
    
    if (!serviceRegistry.isServiceConnected(req.session, serviceType)) {
      return res.status(401).json({ error: `${serviceType} not connected` });
    }

    const connectedServices = serviceRegistry.getConnectedServices(req.session);
    const service = connectedServices[serviceType];

    if (!service) {
      return res.status(404).json({ error: 'Service not found' });
    }

    const tracks = await service.getPlaylistTracks(playlistId);
    
    res.json({
      service: serviceType,
      playlistId,
      tracks: tracks.map(track => ({
        id: track.id,
        name: track.name,
        artist: track.artist,
        album: track.album,
        duration: track.duration,
        isrc: track.isrc
      }))
    });
  } catch (error) {
    console.error(`Error fetching tracks from ${req.params.serviceType}:`, error);
    res.status(500).json({ error: 'Failed to fetch tracks' });
  }
});

/**
 * Validate sync requirements between two services
 */
router.post('/validate-sync', (req, res) => {
  try {
    const { sourceType, destinationType } = req.body;

    if (!sourceType || !destinationType) {
      return res.status(400).json({ 
        error: 'Both sourceType and destinationType are required' 
      });
    }

    const validation = serviceRegistry.validateSyncRequirements(
      req.session, 
      sourceType, 
      destinationType
    );

    if (!validation.success) {
      return res.status(400).json({ 
        valid: false, 
        error: validation.message 
      });
    }

    // Get service capabilities for the UI
    const sourceCapabilities = serviceRegistry.getServiceCapabilities(sourceType);
    const destCapabilities = serviceRegistry.getServiceCapabilities(destinationType);

    res.json({
      valid: true,
      capabilities: {
        supportsISRC: sourceCapabilities?.supportsISRC && destCapabilities?.supportsISRC,
        maxTracks: Math.min(
          sourceCapabilities?.maxPlaylistTracks || Infinity,
          destCapabilities?.maxPlaylistTracks || Infinity
        ),
        sourceService: {
          name: serviceRegistry.serviceTypes[sourceType]?.name,
          capabilities: sourceCapabilities
        },
        destinationService: {
          name: serviceRegistry.serviceTypes[destinationType]?.name,
          capabilities: destCapabilities
        }
      }
    });
  } catch (error) {
    console.error('Error validating sync:', error);
    res.status(500).json({ error: 'Failed to validate sync requirements' });
  }
});

module.exports = router;

File: app/routes/services.py
from fastapi import APIRouter, Request, HTTPException

from app.services.service_registry import ServiceRegistry

router = APIRouter()

@router.get("/")
async def get_services(request: Request):
    """Get all available services with connection status"""
    try:
        connection_status = ServiceRegistry.get_connection_status(request.session)
        connected_users = ServiceRegistry.get_connected_service_users(request.session)
        available_services = ServiceRegistry.get_available_services()

        services = []
        for service in available_services:
            service_data = {
                **service,
                "connected": connection_status.get(service["type"], False),
                "user": connected_users.get(service["type"])
            }
            services.append(service_data)

        return {
            "services": services,
            "totalConnected": sum(1 for connected in connection_status.values() if connected)
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get services: {str(e)}")

@router.get("/connected")
async def get_connected_services(request: Request):
    """Get only connected services"""
    try:
        connection_status = ServiceRegistry.get_connection_status(request.session)
        connected_users = ServiceRegistry.get_connected_service_users(request.session)
        available_services = ServiceRegistry.get_available_services()

        connected_services = []
        for service in available_services:
            if connection_status.get(service["type"], False):
                service_data = {
                    **service,
                    "connected": True,
                    "user": connected_users.get(service["type"])
                }
                connected_services.append(service_data)

        return {
            "services": connected_services,
            "count": len(connected_services)
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get connected services: {str(e)}")

@router.get("/{service_type}")
async def get_service_info(request: Request, service_type: str):
    """Get information about a specific service"""
    try:
        service_info = ServiceRegistry.get_service_info(service_type)
        if not service_info:
            raise HTTPException(status_code=404, detail="Service not found")

        connection_status = ServiceRegistry.get_connection_status(request.session)
        connected_users = ServiceRegistry.get_connected_service_users(request.session)

        return {
            **service_info,
            "connected": connection_status.get(service_type.lower(), False),
            "user": connected_users.get(service_type.lower())
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get service info: {str(e)}")

@router.post("/{service_type}/disconnect")
async def disconnect_service(request: Request, service_type: str):
    """Disconnect from a specific service"""
    try:
        success = ServiceRegistry.disconnect_service(request.session, service_type)

        if not success:
            raise HTTPException(status_code=400, detail="Failed to disconnect service")

        return {
            "success": True,
            "message": f"Disconnected from {service_type}",
            "service": service_type
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to disconnect service: {str(e)}")

@router.get("/{service_type}/capabilities")
async def get_service_capabilities(request: Request, service_type: str):
    """Get capabilities of a specific service"""
    try:
        # Check if service is connected
        connection_status = ServiceRegistry.get_connection_status(request.session)
        if not connection_status.get(service_type.lower()):
            raise HTTPException(status_code=401, detail=f"{service_type} is not connected")

        # Get connected services
        connected_services = ServiceRegistry.get_connected_services(request.session)
        service_instance = connected_services.get(service_type.lower())

        if not service_instance:
            raise HTTPException(status_code=401, detail=f"Failed to get {service_type} service instance")

        capabilities = service_instance.get_capabilities()

        return {
            "service": service_type,
            "capabilities": capabilities
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get service capabilities: {str(e)}")

@router.get("/{service_type}/user")
async def get_service_user(request: Request, service_type: str):
    """Get user information for a specific service"""
    try:
        # Check if service is connected
        connection_status = ServiceRegistry.get_connection_status(request.session)
        if not connection_status.get(service_type.lower()):
            raise HTTPException(status_code=401, detail=f"{service_type} is not connected")

        # Get connected services
        connected_services = ServiceRegistry.get_connected_services(request.session)
        service_instance = connected_services.get(service_type.lower())

        if not service_instance:
            raise HTTPException(status_code=401, detail=f"Failed to get {service_type} service instance")

        try:
            user = service_instance.get_current_user()
            return {
                "service": service_type,
                "user": user
            }
        except Exception as e:
            # Return basic info if detailed user info fails
            connected_users = ServiceRegistry.get_connected_service_users(request.session)
            user = connected_users.get(service_type.lower(), {})

            return {
                "service": service_type,
                "user": user
            }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get service user: {str(e)}")